File: paimon-filesystems/paimon-oss-impl/src/main/java/org/apache/paimon/oss/OSSBlobPresigner.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.oss;

import org.apache.paimon.data.BlobDescriptor;
import org.apache.paimon.fs.Path;
import org.apache.paimon.utils.BlobDescriptorUtils;

import com.aliyun.oss.HttpMethod;
import com.aliyun.oss.OSSClient;
import com.aliyun.oss.OSSException;
import com.aliyun.oss.model.AbortMultipartUploadRequest;
import com.aliyun.oss.model.CompleteMultipartUploadRequest;
import com.aliyun.oss.model.InitiateMultipartUploadRequest;
import com.aliyun.oss.model.InitiateMultipartUploadResult;
import com.aliyun.oss.model.ObjectMetadata;
import com.aliyun.oss.model.PartETag;
import com.aliyun.oss.model.UploadPartCopyRequest;
import com.aliyun.oss.model.UploadPartCopyResult;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.net.URI;
import java.net.URL;
import java.net.URLConnection;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/** Materializes blob ranges and creates OSS presigned URLs. */
public final class OSSBlobPresigner {

    private static final String BLOB_FINGERPRINT_METADATA = "paimon-blob-descriptor-sha256";
    private static final long BLOB_COPY_MIN_PART_SIZE = 100L * 1024 * 1024;
    private static final long MAX_MULTIPART_UPLOAD_PARTS = 10_000;
    private static final char[] HEX_CHARS = "0123456789abcdef".toCharArray();

    private OSSBlobPresigner() {}

    public static String create(
            OSSClient client,
            Path tableRoot,
            BlobDescriptor descriptor,
            String extension,
            Duration validity)
            throws IOException {
        BlobDescriptorUtils.validateTableRoot(tableRoot, descriptor);
        if (validity == null
                || validity.isZero()
                || validity.isNegative()
                || validity.getNano() != 0) {
            throw new IOException("Blob presigned URL validity must be positive whole seconds.");
        }

        String normalizedExtension = normalizeExtension(extension);
        String contentType = contentType(normalizedExtension);
        try {
            URI source = new Path(descriptor.uri()).toUri();
            String bucket = source.getAuthority();
            String sourceKey = objectKey(source);
            String fingerprint = sha256Hex(descriptor.serialize());
            int parentEnd = sourceKey.lastIndexOf('/') + 1;
            String targetKey =
                    sourceKey.substring(0, parentEnd)
                            + "_bloburl_"
                            + fingerprint
                            + "."
                            + normalizedExtension;

            ObjectMetadata target = headObjectIfExists(client, bucket, targetKey);
            if (!matches(target, descriptor.length(), contentType, fingerprint)) {
                ObjectMetadata sourceMetadata = client.headObject(bucket, sourceKey);
                validateRange(descriptor, sourceMetadata.getContentLength());
                materialize(
                        client, bucket, sourceKey, targetKey, descriptor, contentType, fingerprint);
                target = client.headObject(bucket, targetKey);
                if (!matches(target, descriptor.length(), contentType, fingerprint)) {
                    throw new IOException(
                            "Materialized blob object metadata does not match descriptor.");
                }
            }

            URL url =
                    client.generatePresignedUrl(
                            bucket,
                            targetKey,
                            Date.from(Instant.now().plus(validity)),
                            HttpMethod.GET);
            validatePresignedUrl(client, url, bucket, targetKey);
            return url.toString();
        } catch (IOException e) {
            throw e;
        } catch (Exception e) {
            throw new IOException("Failed to create blob presigned URL.", e);
        }
    }

    private static void materialize(
            OSSClient client,
            String bucket,
            String sourceKey,
            String targetKey,
            BlobDescriptor descriptor,
            String contentType,
            String fingerprint)
            throws Exception {
        ObjectMetadata metadata = new ObjectMetadata();
        metadata.setContentType(contentType);
        metadata.addUserMetadata(BLOB_FINGERPRINT_METADATA, fingerprint);
        if (descriptor.length() == 0) {
            metadata.setContentLength(0);
            client.putObject(bucket, targetKey, new ByteArrayInputStream(new byte[0]), metadata);
            return;
        }

        String uploadId = null;
        try {
            InitiateMultipartUploadResult initiated =
                    client.initiateMultipartUpload(
                            new InitiateMultipartUploadRequest(bucket, targetKey, metadata));
            uploadId = initiated.getUploadId();
            long partSize =
                    Math.max(
                            BLOB_COPY_MIN_PART_SIZE,
                            descriptor.length() / MAX_MULTIPART_UPLOAD_PARTS + 1);
            List<PartETag> parts = new ArrayList<>();
            long copied = 0;
            int partNumber = 1;
            while (copied < descriptor.length()) {
                long size = Math.min(partSize, descriptor.length() - copied);
                UploadPartCopyResult result =
                        client.uploadPartCopy(
                                new UploadPartCopyRequest(
                                        bucket,
                                        sourceKey,
                                        bucket,
                                        targetKey,
                                        uploadId,
                                        partNumber,
                                        descriptor.offset() + copied,
                                        size));
                parts.add(new PartETag(partNumber, result.getETag()));
                copied += size;
                partNumber++;
            }
            client.completeMultipartUpload(
                    new CompleteMultipartUploadRequest(bucket, targetKey, uploadId, parts));
            uploadId = null;
        } catch (Exception e) {
            if (uploadId != null) {
                try {
                    client.abortMultipartUpload(
                            new AbortMultipartUploadRequest(bucket, targetKey, uploadId));
                } catch (Exception abortException) {
                    e.addSuppressed(abortException);
                }
            }
            throw e;
        }
    }

    private static ObjectMetadata headObjectIfExists(OSSClient client, String bucket, String key) {
        try {
            return client.headObject(bucket, key);
        } catch (OSSException e) {
            if ("NoSuchKey".equals(e.getErrorCode()) || "NoSuchObject".equals(e.getErrorCode())) {
                return null;
            }
            throw e;
        }
    }

    private static boolean matches(
            ObjectMetadata metadata, long length, String contentType, String fingerprint) {
        return metadata != null
                && metadata.getContentLength() == length
                && contentType.equals(metadata.getContentType())
                && fingerprint.equals(metadata.getUserMetadata().get(BLOB_FINGERPRINT_METADATA));
    }

    private static void validateRange(BlobDescriptor descriptor, long sourceLength)
            throws IOException {
        if (descriptor.offset() < 0
                || descriptor.length() < 0
                || descriptor.offset() > sourceLength
                || descriptor.length() > sourceLength - descriptor.offset()) {
            throw new IOException("Blob descriptor range is outside the source object.");
        }
    }

    private static String normalizeExtension(String extension) throws IOException {
        if (extension == null || extension.isEmpty()) {
            throw new IOException("Blob file extension must contain only ASCII letters or digits.");
        }
        for (int i = 0; i < extension.length(); i++) {
            char c = extension.charAt(i);
            if (!(c >= 'a' && c <= 'z') && !(c >= 'A' && c <= 'Z') && !(c >= '0' && c <= '9')) {
                throw new IOException(
                        "Blob file extension must contain only ASCII letters or digits.");
            }
        }
        return extension.toLowerCase(Locale.ROOT);
    }

    private static String contentType(String extension) throws IOException {
        String contentType = URLConnection.guessContentTypeFromName("file." + extension);
        if (contentType == null || "application/octet-stream".equals(contentType)) {
            throw new IOException("Unknown MIME type for blob file extension: " + extension);
        }
        return contentType;
    }

    private static String objectKey(URI uri) throws IOException {
        String path = uri.getPath();
        if (path == null || !path.startsWith("/") || path.length() == 1) {
            throw new IOException("Blob descriptor URI must contain an OSS object key.");
        }
        return path.substring(1);
    }

    private static String sha256Hex(byte[] bytes) {
        try {
            byte[] hash = MessageDigest.getInstance("SHA-256").digest(bytes);
            char[] chars = new char[hash.length * 2];
            for (int i = 0; i < hash.length; i++) {
                chars[i * 2] = HEX_CHARS[(hash[i] >> 4) & 0x0f];
                chars[i * 2 + 1] = HEX_CHARS[hash[i] & 0x0f];
            }
            return new String(chars);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available.", e);
        }
    }

    private static void validatePresignedUrl(
            OSSClient client, URL url, String bucket, String targetKey) throws Exception {
        URI endpoint = client.getEndpoint();
        String expectedHost = bucket + "." + endpoint.getHost();
        if (!"https".equalsIgnoreCase(endpoint.getScheme())
                || !"https".equalsIgnoreCase(url.getProtocol())
                || !expectedHost.equalsIgnoreCase(url.getHost())
                || !("/" + targetKey).equals(url.toURI().getPath())) {
            throw new IOException("OSS client generated a presigned URL for an invalid target.");
        }
    }
}

File: paimon-python/pypaimon/filesystem/oss_blob_presigner.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""Materialize OSS BLOB ranges and create temporary GET URLs."""

import hashlib
import logging
import posixpath
import re
from datetime import timedelta
from urllib.parse import unquote, urlparse, urlsplit


_LOGGER = logging.getLogger(__name__)

_BLOB_FINGERPRINT_METADATA = "paimon-blob-descriptor-sha256"
_BLOB_FINGERPRINT_HEADER = "x-oss-meta-" + _BLOB_FINGERPRINT_METADATA
_BLOB_CONTENT_TYPE = "application/octet-stream"
_BLOB_COPY_MIN_PART_SIZE = 100 * 1024 * 1024
_MAX_MULTIPART_UPLOAD_PARTS = 10_000


def create_presigned_url(
        bucket, table_root, descriptor, validity, sse_headers=None,
        has_security_token=False) -> str:
    """Create a presigned URL with the same object layout as Java Paimon."""
    endpoint = urlsplit(bucket.endpoint)
    if endpoint.scheme.lower() != "https" or not endpoint.hostname:
        raise ValueError("OSS BLOB presigning requires an HTTPS endpoint.")
    validity_seconds = _validity_seconds(validity, has_security_token)
    source = _validate_table_root(table_root, descriptor)
    # Keep the compatibility URI rules in OssFileIO, which also owns bucket
    # extraction for oss://AK:SK@endpoint/bucket/object-key URIs.
    from pypaimon.filesystem.oss_file_io import OssFileIO
    source_key = OssFileIO._extract_oss_object_key(source.geturl())
    if not source_key:
        raise ValueError("Blob descriptor URI must contain an OSS object key.")
    sse_headers = dict(sse_headers or {})
    fingerprint = hashlib.sha256(descriptor.serialize()).hexdigest()
    parent_end = source_key.rfind('/') + 1
    target_key = source_key[:parent_end] + "_bloburl_" + fingerprint

    try:
        target = _head_object_if_exists(bucket, target_key)
        if not _matches(
                target, descriptor.length, fingerprint, sse_headers):
            source_metadata = bucket.head_object(source_key)
            _validate_range(descriptor, source_metadata.content_length)
            _materialize(
                bucket,
                source_key,
                target_key,
                descriptor,
                fingerprint,
                sse_headers,
            )
            target = bucket.head_object(target_key)
            if not _matches(
                    target, descriptor.length, fingerprint, sse_headers):
                raise OSError(
                    "Materialized blob object metadata does not match "
                    "descriptor.")

        url = bucket.sign_url(
            'GET', target_key, validity_seconds, slash_safe=True)
        _validate_presigned_url(bucket, url, target_key)
        return url
    except (OSError, ValueError):
        raise
    except Exception as error:
        raise OSError("Failed to create blob presigned URL.") from error


def _validity_seconds(validity, has_security_token=False) -> int:
    if not isinstance(validity, timedelta):
        raise TypeError(
            "Blob presigned URL validity must be datetime.timedelta.")
    if validity <= timedelta(0) or validity.microseconds != 0:
        raise ValueError(
            "Blob presigned URL validity must be positive whole seconds.")
    seconds = validity.days * 24 * 60 * 60 + validity.seconds
    maximum = 43200 if has_security_token else 604800
    if seconds > maximum:
        raise ValueError(
            "OSS V4 presigned URL validity must not exceed {} seconds."
            .format(maximum))
    return seconds


def _validate_table_root(table_root, descriptor):
    root = urlparse(str(table_root))
    blob = urlparse(descriptor.uri)
    root_path = _normalize_path(root.path)
    blob_path = _normalize_path(blob.path)
    if ((root.scheme or '').lower() != (blob.scheme or '').lower()
            or root.netloc != blob.netloc):
        raise ValueError(
            "Blob descriptor URI must use the same scheme and authority "
            "as table root.")
    child_prefix = root_path if root_path.endswith('/') else root_path + '/'
    if not blob_path.startswith(child_prefix):
        raise ValueError("Blob descriptor URI must be under table root.")
    return blob._replace(path=blob_path)


def _normalize_path(path):
    normalized = posixpath.normpath(re.sub('/+', '/', path or '/'))
    if path.startswith('/') and not normalized.startswith('/'):
        normalized = '/' + normalized
    return normalized


def _head_object_if_exists(bucket, key):
    try:
        return bucket.head_object(key)
    except Exception as error:
        if getattr(error, 'code', None) in ('NoSuchKey', 'NoSuchObject'):
            return None
        raise


def _matches(metadata, length, fingerprint, sse_headers) -> bool:
    if metadata is None:
        return False
    headers = getattr(metadata, 'headers', {}) or {}
    metadata_fingerprint = next(
        (value for key, value in headers.items()
         if key.lower() == _BLOB_FINGERPRINT_HEADER),
        None,
    )
    return (
        metadata.content_length == length
        and metadata.content_type == _BLOB_CONTENT_TYPE
        and metadata_fingerprint == fingerprint
        and _matches_sse_headers(headers, sse_headers)
    )


def _matches_sse_headers(actual_headers, expected_headers) -> bool:
    if not expected_headers:
        return True
    actual = {
        str(key).lower(): value
        for key, value in actual_headers.items()
    }
    case_insensitive_values = {
        'x-oss-server-side-encryption',
        'x-oss-server-side-data-encryption',
    }
    for key, expected in expected_headers.items():
        normalized_key = str(key).lower()
        found = actual.get(normalized_key)
        if found is None:
            return False
        if normalized_key in case_insensitive_values:
            if str(found).upper() != str(expected).upper():
                return False
        elif found != expected:
            return False
    return True


def _validate_range(descriptor, source_length):
    if (descriptor.offset < 0
            or descriptor.length < 0
            or descriptor.offset > source_length
            or descriptor.length > source_length - descriptor.offset):
        raise ValueError("Blob descriptor range is outside the source object.")


def _materialize(
        bucket,
        source_key,
        target_key,
        descriptor,
        fingerprint,
        sse_headers):
    import oss2

    headers = dict(sse_headers)
    headers.update({
        'Content-Type': _BLOB_CONTENT_TYPE,
        _BLOB_FINGERPRINT_HEADER: fingerprint,
    })
    if descriptor.length == 0:
        bucket.put_object(target_key, b'', headers=headers)
        return

    upload_id = None
    try:
        initiated = bucket.init_multipart_upload(target_key, headers=headers)
        upload_id = initiated.upload_id
        part_size = max(
            _BLOB_COPY_MIN_PART_SIZE,
            descriptor.length // _MAX_MULTIPART_UPLOAD_PARTS + 1,
        )
        parts = []
        copied = 0
        part_number = 1
        while copied < descriptor.length:
            size = min(part_size, descriptor.length - copied)
            result = bucket.upload_part_copy(
                bucket.bucket_name,
                source_key,
                (descriptor.offset + copied,
                 descriptor.offset + copied + size - 1),
                target_key,
                upload_id,
                part_number,
            )
            parts.append(oss2.models.PartInfo(part_number, result.etag))
            copied += size
            part_number += 1
        bucket.complete_multipart_upload(target_key, upload_id, parts)
        upload_id = None
    except Exception:
        if upload_id is not None:
            try:
                bucket.abort_multipart_upload(target_key, upload_id)
            except Exception:
                _LOGGER.warning(
                    "Failed to abort OSS BLOB multipart upload %s for %s",
                    upload_id, target_key, exc_info=True)
        raise


def _validate_presigned_url(bucket, url, target_key):
    endpoint = urlsplit(bucket.endpoint)
    actual = urlsplit(url)
    expected_host = bucket.bucket_name + '.' + (endpoint.hostname or '')
    if (endpoint.scheme.lower() != 'https'
            or actual.scheme.lower() != 'https'
            or (actual.hostname or '').lower() != expected_host.lower()
            or unquote(actual.path) != '/' + target_key):
        raise OSError(
            "OSS client generated a presigned URL for an invalid target.")
